# RNAX
A computational pipeline for variant calling and expression quantification of bulk RNA-sequencing data.

## Pre-requisites
* familiarity with Linux Command-line system
* familiarity with Nextflow (optional)
* installation of Nextflow DSL2 (see Installation section below)
* HPC usage
    * installation of Singularity and Slurm
    * you - the user - probably do not need to do this yourself!
* local usage
    * installation of Singularity

## Installation
* in your `home` directory, install java
    1. `wget https://download.oracle.com/java/17/archive/jdk-17.0.7_linux-x64_bin.tar.gz`
    2. `tar -xf jdk-17.0.7_linux-x64_bin.tar.gz`
    3. `mkdir jdk`
    4. `mv jdk-17.0.7 jdk/`
    5. add the following to your `~/.bashrc`: 

    ```bash
    # global environment variables
    export JAVA_HOME="/c4/home/${USER}/jdk/jdk-17.0.7"
    export NXF_JAVA_HOME="/c4/home/${USER}/jdk/jdk-17.0.7"
    export PATH=$JAVA_HOME/bin:$PATH
    ```
* in your `home` directory, install nextflow
    1. If `bin` does not exist: `mkdir bin`
    2. `cd bin`
    3. `wget -qO- https://get.nextflow.io | bash`
    4. `chmod +x nextflow`
* clone code repository from GitHub
```bash
git clone https://github.com/UCSF-DSCOLAB/data_processing_pipelines.git
```
* go into the bulk RNAseq directory
```bash
cd data_processing_pipelines/bulk_RNAseq
```

## Configuration
* Create a sample sheet (CSV) specifying each sample, locations of their sequencing reads (FASTQ), and whether they are single-end or paired-end reads. The figure below is an example of a correctly formatted sample sheet:
* Avoid creating the csv using Microsoft Excel because it can add weird symbols (eg. <U+FEFF>) to the beginning of the file causing issues for the pipeline
![sample_sheet](docs/figs/sample_sheet_example.png)
* If one or more of your samples have multiple FASTQ reads (i.e. multiple lanes), the sample sheet should contain a row for each lane with an identical name under the `sample` column
* Open `config/user_directories.config` using your favorite text editor, and specify the following parameters before saving the file:
    * `results_directory`: for storing final results; this should be in a location that you have read-write access to
    * `input_sample_sheet`: sample sheet name. THIS FILE MUST BE INSIDE `results_directory`
    * `tmp_dir`: for storing temporary files generated by the different tools; it should be the same as your main temporary directory e.g. `/c4/scratch/[username]`

### Usage (on C4 compute nodes)
* Run the DSL2 pipeline using Slurm and Singularity
    * NOTE: you must run the following command inside `data_processing_pipelines/bulk_RNASeq/`
```bash
sbatch ./run_pipeline.sh -profile hpc
```

#### Additional Functionality
* To resume a pipeline run on Slurm 
```bash
sbatch ./run_pipeline.sh -profile hpc -resume
```

### Local Usage
* Run the DSL2 pipeline using Anaconda
```bash
nextflow run bulk_rna_seq.nf -c config/base.config -w your/tmp/directory -profile local
```
#### Additional Functionality
* To resume a pipeline run on local
```bash
nextflow run bulk_rna_seq.nf -c config/base.config -w your/tmp/directory -profile local -resume
```

### Troubleshooting
* If a run fails, examine log ouputs to identify and resolve issues before resuming run
* Common cause of failure includes insufficent memory allocated for jobs (solution: for the process that ran out of memory, adjust the amount of computational resource allocated in `container.config` for _HPC Usage_ or `conda.config` for _Local Usage_)

## Authors
Emily Flynn

Al Latif

Amadeo Mazzara

Tammie Tam

Daniel Bunis

Walter Eckalbar